sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionService.java

```diff
@@ -362,6 +368,19 @@ default InputT createInput(Pipeline p, Map<String, PCollection<?>> inputs) {
   }
 
   private @MonotonicNonNull Map<String, TransformProvider> registeredTransforms;
+  private final PipelineOptions pipelineOptions;
+
+  public ExpansionService() {
+    this(new String[] {});
+  }
+
+  public ExpansionService(String[] args) {
+    this(PipelineOptionsFactory.fromArgs(args).create());
+  }
+
+  public ExpansionService(PipelineOptions opts) {
```

Comment:
I think we should implement some sort of a validation to limit the supported options in the first version of the PR (this can be based on annotations as you mentioned in the mailing list).

Reply:
That is more complex than I expected. I actually don't know how to do that. The only "legal" option seems to be serializing the options to JSON and then parse them back. That is more complicated than I would expect. The current approach to validation is that it validates only requirements - and the validation is called in `as(Class)`. I could add a new validation type there, but it would require changing public API of PipelineOptionsFactory - probably the best would be to add annotation of allowed scopes to getter methods. The PipelineOptionsFactory.fromArgs(args) would have to be extended to accept the set of allowed scopes (default all scopes). This seems to be a non-trivial change. I'll try to do a sketch, but I think this would really deserve new JIRA and a new PR. We might try to get that to 2.32.0 as well.

Reply:
I have a sketch, but I'm missing access to the annotations. https://github.com/je-ik/beam/commit/a78ed9b87270f131c5eee4bdcd6f9de772d6aef0
The AllowedScopes might be renamed (maybe SupportedScopes), but as I said, I'd prefer to push this to different PR to have the discussion.

Reply:
Till we have proper validation in place, can we just explicitly create a second PipelineOptions with only the set of options that we wish to support ? (to make sure that arbitrary options that may or may not be supported do not get passed to the external transform expansion).

Reply:
We can do that, but the question is which case exactly do we want to solve. My initial feeling was that we want to throw exception when user sets option that is not supported (that is, might not be respected by the ExpansionService). Extracting 'supported' features will not help this - unsupported options will still be silently ignored by the service, pretty much the same way as we do not extract it.